protocols/autonat/src/v2/server/handler/dial_request.rs

```diff
@@ -0,0 +1,349 @@
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use rand_core::RngCore;
+
+use crate::v2::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    protocol::{Coder, DialDataRequest, DialRequest, DialResponse, Request, Response},
+    server::behaviour::StatusUpdate,
+    Nonce, DIAL_REQUEST_PROTOCOL,
+};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum DialBackStatus {
+    /// Failure during dial
+    DialErr,
+    /// Failure during dial back
+    DialBackErr,
+}
+
+#[derive(Debug)]
+pub struct DialBackCommand {
+    pub(crate) addr: Multiaddr,
+    pub(crate) nonce: Nonce,
+    pub(crate) back_channel: oneshot::Sender<Result<(), DialBackStatus>>,
+}
+
+pub struct Handler<R> {
+    client_id: PeerId,
+    observed_multiaddr: Multiaddr,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    dial_back_cmd_receiver: mpsc::Receiver<DialBackCommand>,
+    status_update_sender: mpsc::Sender<StatusUpdate>,
+    status_update_receiver: mpsc::Receiver<StatusUpdate>,
+    inbound: FuturesSet<Result<(), io::Error>>,
+    rng: R,
+}
+
+impl<R> Handler<R>
+where
+    R: RngCore,
+{
+    pub(crate) fn new(client_id: PeerId, observed_multiaddr: Multiaddr, rng: R) -> Self {
+        let (dial_back_cmd_sender, dial_back_cmd_receiver) = mpsc::channel(10);
+        let (status_update_sender, status_update_receiver) = mpsc::channel(10);
+        Self {
+            client_id,
+            observed_multiaddr,
+            dial_back_cmd_sender,
+            dial_back_cmd_receiver,
+            status_update_sender,
+            status_update_receiver,
+            inbound: FuturesSet::new(Duration::from_secs(10), 10),
+            rng,
+        }
+    }
+}
+
+impl<R> ConnectionHandler for Handler<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    type FromBehaviour = ();
+
+    type ToBehaviour = Either<io::Result<DialBackCommand>, StatusUpdate>;
+
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
+
+    type OutboundProtocol = DeniedUpgrade;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(ReadyUpgrade::new(DIAL_REQUEST_PROTOCOL), ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        match self.inbound.poll_unpin(cx) {
+            Poll::Ready(Ok(Err(e))) => {
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Either::Left(Err(
+                    e,
+                ))));
+            }
+            Poll::Ready(Err(e)) => {
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Either::Left(Err(
+                    io::Error::new(io::ErrorKind::TimedOut, e),
+                ))));
+            }
```

Comment:
Is it worth sending these errors to the behaviour? It has no context for these errors so all it can do is log them. Might as well log them on this layer already then? Makes the API between those two modules simpler.

Reply:
you are right, that seems like a better solution.